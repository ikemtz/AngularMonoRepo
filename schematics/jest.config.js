module.exports = {
  testMatch: ["**/+(*.)+(spec|test).+(ts)?(x)"],
  transform: {
    "^.+\\.(ts|js|html)$": "ts-jest",
  },
  testPathIgnorePatterns: ["/node_modules/", "/dist/"],
  moduleFileExtensions: ["ts", "js", "html"],
  collectCoverage: true,
  coverageReporters: ["html", "lcov", "cobertura"],
  reporters: [
    "default",
    [
      "jest-junit",
      {
        outputDirectory: ".",
        outputName: "./junit.schematics.xml",
      },
    ],
  ],
  moduleNameMapper: {},
};
