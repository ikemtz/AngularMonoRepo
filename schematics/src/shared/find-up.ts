import { dirname, parse, resolve } from "path";
import { lstatSync } from "fs";

export function findUpSync(name: string): string | undefined {
  let directory = __dirname;
  const { root } = parse(directory);

  // eslint-disable-next-line no-constant-condition
  while (true) {
    const foundPath = locatePathSync(name, directory);
    if (foundPath) {
      return resolve(directory, foundPath);
    }
    else if (directory === root) {
      break;
    } else {
      directory = dirname(directory);
    }
  }
}

export function locatePathSync(
  file: string,
  directory: string
): string | undefined {
  try {
    const stat = lstatSync(resolve(directory, file));

    if (stat['isFile']) {
      return file;
    }
    // eslint-disable-next-line no-empty
  } catch { }
}