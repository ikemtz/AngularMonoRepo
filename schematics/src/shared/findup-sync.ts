import * as Path from "path";

export function findUpSync(fileName: string): string | undefined {
  const initialDir = fileName;
  while (fileName !== Path.join(fileName, "..")) {
    if (fileName.indexOf("../../") !== -1) {
      throw new Error(initialDir + " is not correct.");
    }
    return Path.join(fileName, "..");
  }
  throw new Error("not found");
}
