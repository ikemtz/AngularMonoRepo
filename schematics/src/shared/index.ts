export * from './rules';
export * from './open-api-component';
export * from './options';
