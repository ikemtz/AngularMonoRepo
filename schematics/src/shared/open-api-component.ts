export interface OpenApiComponent {
  properties: Properties;
  required?: string[];
}
export interface Properties {
  [key: string]: PropertyInfo;
}
export interface PropertyInfo {
  name?: string;
  enum?: [string | number];
  type: 'string' | 'boolean' | 'array' | 'integer' | 'number' | 'date';
  format?: 'date-time' | 'uuid' | 'int32' | 'double';
  nullable?: boolean;
  readOnly?: boolean;
  required: boolean;
  htmlInputType: 'text' | 'number' | 'checkbox' | 'date';
  snakeCaseName: string;
  maxLength: number;
  testFactoryValue: string;
}
