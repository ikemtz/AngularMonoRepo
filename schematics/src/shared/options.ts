import { PropertyInfo } from './open-api-component';

export interface IOptions {
  /** Entity Name (Must match schema name in json) */
  name: string;
  /** Parent Entity Name (Must match schema name in json) */
  parentName?: string;
  /** NGRX Store Feature Name */
  storeName?: string;
  /** Angular application Prefix */
  appPrefix: string;
  /** Folder to place generated files */
  path: string;
  /** Optional: Url of remote OpenApi JSON document (If this is not provided, then you must specify openApiJsonFileName) */
  openApiJsonUrl?: string;
  /** Optional: File path of local OpenApi JSON document (If this is not provided, then you must specify openApiJsonUrl) */
  openApiJsonFileName?: string;
  /** Optional: This is the property that will used for initial sorting on list and auto focus on crud schematics */
  modelFolderLocation?: string;
  /** Optional: Location of model files relative to the module/component generated root */

  firstProperty?: PropertyInfo;
  /** Ignore: This is for internal schematic use only  */
  swaggerProperties: PropertyInfo[];
  /** Ignore: This is for internal schematic use only  */
  swaggerObjectProperties?: PropertyInfo[];
  /** Ignore: This is for internal schematic use only  */
  hasDates?: boolean;
  hasObjects?: boolean;
  hasNullableDates?: boolean;
}
