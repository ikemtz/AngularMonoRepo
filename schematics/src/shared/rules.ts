import {
  apply,
  applyTemplates,
  chain,
  DirEntry,
  mergeWith,
  move,
  Rule,
  SchematicContext,
  Tree,
  url,
} from '@angular-devkit/schematics';
import { strings, normalize } from '@angular-devkit/core';
import fetch from 'node-fetch';
import { PropertyInfo, OpenApiComponent } from './open-api-component';
import * as pluralize from 'pluralize';
import { Observable, from, of } from 'rxjs';
import { map, switchMap } from 'rxjs/operators';
import { IOptions } from './options';
import _ = require('lodash');
import { TslintFixTask } from '@angular-devkit/schematics/tasks';
import * as fs from 'fs';
import * as findUp from 'find-up';
import * as https from 'https';
import * as http from 'http';

export function getSwaggerDoc(options: IOptions): Rule {
  return (host: Tree, context: SchematicContext) => {
    let jsonDoc: Observable<any> | null = null;
    if (options.openApiJsonFileName) {
      const fileName = getFileNames(options.openApiJsonFileName);
      context.logger.info(`Getting Swagger Document @${fileName}`);
      if (!fileName) {
        throw new Error(`File not found: ${options.openApiJsonFileName}`);
      }
      const response = fs.readFileSync(fileName);
      const apiDoc = JSON.parse(response.toString());
      jsonDoc = of(apiDoc);
    } else if (options.openApiJsonUrl) {
      context.logger.info(`Getting Swagger Document @${options.openApiJsonUrl}`);
      const httpsAgent = options.openApiJsonUrl.toLowerCase().startsWith("https") ? new https.Agent({
        rejectUnauthorized: false,
      }) : new http.Agent();
      jsonDoc = from(fetch(options.openApiJsonUrl, {
        agent: httpsAgent
      })).pipe(
        switchMap(resp => from(resp.json())));
    }
    if (jsonDoc) {
      return jsonDoc.pipe(map(data => processOpenApiDoc(data, options, host)));
    }
    return of(host);

  };
}

function getFileNames(openApiJsonFileName: string) {
  if (fs.existsSync(`${__dirname}/${openApiJsonFileName}`)) {
    return `${__dirname}/${openApiJsonFileName}`;
  }
  return findUp.sync(openApiJsonFileName);
}

export function processOpenApiDoc(data: any, schema: IOptions, host: Tree) {
  const openApiComonent = data.components.schemas[strings.classify(schema.name)] as OpenApiComponent;
  if (!openApiComonent) {
    throw new Error(`OpenApi Component not found in swagger doc: ${schema.name}`);
  }
  const properties = openApiComonent.properties as {
    [key: string]: PropertyInfo;
  };
  const filteredProperties: PropertyInfo[] = [];
  const excludedFields = ['createdBy', 'createdOnUtc', 'tenantId', 'updatedBy', 'updatedOnUtc'];
  schema.hasDates = false;
  for (const propertyKey in properties) {
    if (excludedFields.indexOf(propertyKey) < 0 && properties[propertyKey].type !== 'array') {
      const property =
        mapPropertyAttributes(schema, properties[propertyKey], {
          ...properties[propertyKey],
          name: propertyKey,
          required: (openApiComonent.required ?? []).indexOf(propertyKey) > -1,
        });
      filteredProperties.push(property);
    }
  }
  schema.swaggerProperties = filteredProperties;
  return host;
}

function mapPropertyAttributes(options: IOptions, source: PropertyInfo, dest: any) {

  if (source.type === 'number' || source.type === 'integer') {
    dest.htmlInputType = 'number';
    dest.filterExpression = 'numeric';
    dest.testFactoryValue = '0';
  } else if (source.type === 'boolean') {
    dest.htmlInputType = 'checkbox';
    dest.filterExpression = 'boolean';
    dest.testFactoryValue = 'true';
  } else if (source.format === 'date-time') {
    dest.htmlInputType = 'date';
    dest.filterExpression = 'date';
    dest.testFactoryValue = 'new Date()';
    options.hasDates = true;
  } else {
    dest.htmlInputType = 'text';
    dest.filterExpression = 'text';
    dest.testFactoryValue = `'${dest.maxLength ?
      _.snakeCase(dest.name).toUpperCase().substring(0, dest.maxLength) : _.snakeCase(dest.name).toUpperCase()}'`;
  }
  dest.snakeCaseName = _.snakeCase(dest.name);
  dest.startCaseName = _.startCase(dest.name);
  return dest;
}

export function generateFiles(schema: IOptions, templateType: 'list' | 'crud' | 'module'): Rule {
  return (tree: Tree, context: SchematicContext) => {
    context.logger.info(`Running ${strings.capitalize(templateType)} Schematic`);

    const singularizedName = pluralize(schema.name, 1);
    const pluralizedName = pluralize(schema.name, 2);
    const singularizedStoreName = schema.storeName ? pluralize(schema.storeName, 1) : '';
    const pluralizedStoreName = schema.storeName ? pluralize(schema.storeName, 2) : '';

    const folderName = normalize(`${schema.path}/${strings.dasherize(pluralizedName)}-${templateType}`);
    const templateSource = apply(url('./files'), [
      applyTemplates({
        ...strings,
        ...schema,
        pluralizedName,
        singularizedName,
        pluralizedStoreName,
        singularizedStoreName,
      }),
      move(folderName),
    ]);
    return chain([mergeWith(templateSource)])(tree, context);
  };
}

export function runLint(schema: IOptions): Rule {
  return (tree: Tree, context: SchematicContext) => {
    const dir: DirEntry | null = tree.getDir(schema.path.substr(0, schema.path.lastIndexOf('/')));
    const files = tree.actions.reduce((acc: Set<string>, action) => {
      const path = action.path.substr(1); // Remove the starting '/'.
      if (path.endsWith('.ts') && dir && action.path.startsWith(dir.path)) {
        acc.add(path);
      }
      return acc;
    }, new Set<string>());
    context.addTask(
      new TslintFixTask({
        ignoreErrors: false,
        tsConfigPath: 'tsconfig.json',
        files: Array.from(files),
      }),
    );
    return tree;
  };
}
