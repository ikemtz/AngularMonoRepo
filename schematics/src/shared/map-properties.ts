import * as pluralize from 'pluralize';
import { PropertyInfo, OpenApiComponent, OpenApiDocument } from './open-api-component';
import { IOptions } from './options';
import { mapPropertyAttributes } from './rules';

const excludedFields = ['createdBy', 'createdOnUtc', 'tenantId', 'updatedBy', 'updatedOnUtc'];

export function mapProperties(properties: { [key: string]: PropertyInfo; }, options: IOptions, openApiComponent: OpenApiComponent, openApiDoc: OpenApiDocument) {
  let firstProperty: PropertyInfo | undefined;
  const filteredProperties: PropertyInfo[] = [];
  for (const propertyKey in properties) {
    const originalProperty = properties[propertyKey];
    if (excludedFields.indexOf(propertyKey) < 0 && originalProperty.type !== 'array') {
      const property = mapPropertyAttributes(options, originalProperty, {
        ...properties[propertyKey],
        name: propertyKey,
        required: (openApiComponent.required ?? []).indexOf(propertyKey) > -1,
      });

      mapReferencedProperties(property, propertyKey, openApiComponent, openApiDoc);
      if (!firstProperty && propertyKey !== 'id') {
        firstProperty = property;
      }
      filteredProperties.push(property);
    }
  }
  return { firstProperty, filteredProperties };
}
function mapReferencedProperties(property: PropertyInfo, propertyKey: string, openApiComponent: OpenApiComponent, openApiDoc: OpenApiDocument) {
  const prop = openApiComponent.properties[propertyKey];
  const propTypeNames = prop['$ref']?.split('/');
  if (propTypeNames) {
    const typeName: string = propTypeNames.pop() || '';
    property.propertyTypeName = typeName;
    property.pluralizedPropertyTypeName = pluralize.plural(typeName);
    property.singularizedPropertyTypeName = pluralize.singular(typeName);
    const refComponent = openApiDoc.components.schemas[typeName];
    property.enum = refComponent.enum;
    if (property.enum) {
      const value = property.enum[0].toString().split(' ').pop();
      property.testFactoryValue = `${property.pluralizedPropertyTypeName}.${value}`
    }
  }
}

