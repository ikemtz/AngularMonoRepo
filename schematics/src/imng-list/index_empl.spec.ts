import { Tree } from '@angular-devkit/schematics';
import { SchematicTestRunner, UnitTestTree } from '@angular-devkit/schematics/testing';
import * as path from 'path';
import { IOptions } from '../shared';
import { readFirst } from '@nrwl/angular/testing';
import { classify } from '@angular-devkit/core/src/utils/strings';
import * as pluralize from 'pluralize';

const collectionPath = path.join(__dirname, '../collection.json');


describe('imng-list', () => {
  it('works', async done => {
    try {
      const runner = new SchematicTestRunner('schematics', collectionPath);
      const options: IOptions = {
        name: 'employee',
        openApiJsonUrl: 'https://im-wa-empo-nrcrn.azurewebsites.net/swagger/v1/swagger.json',
        path: './test',
        swaggerProperties: [],
        storeName: 'employees',
        appPrefix: 'nrcrn'
      };
      const tree: UnitTestTree = await readFirst(runner.runSchematicAsync('imng-list', options, Tree.empty()) as any);

      expect(tree.files).toEqual([
        `/test/${pluralize(options.name)}-list/index.ts`,
        `/test/${pluralize(options.name)}-list/list.component.html`,
        `/test/${pluralize(options.name)}-list/list.component.scss`,
        `/test/${pluralize(options.name)}-list/list.component.spec.ts`,
        `/test/${pluralize(options.name)}-list/list.component.ts`,
        `/test/${pluralize(options.name)}-list/list.facade.spec.ts`,
        `/test/${pluralize(options.name)}-list/list.facade.ts`,
      ]);

      const htmlFile = tree.get(`/test/${pluralize(options.name)}-list/list.component.html`);
      let content = htmlFile?.content.toString();
      expect(content).toContain('[field]="props.ADDRESS_LINE_1"');
      expect(content).toContain(`<${options.appPrefix}-${options.name}-add `);
      expect(content).toContain(`<${options.appPrefix}-${options.name}-edit `);

      const componentFile = tree.get(`/test/${pluralize(options.name)}-list/list.component.ts`);
      content = componentFile?.content.toString();
      expect(content).toContain(`${classify(options.name)}Properties.ADDRESS_LINE_1,`);
      expect(content).toContain(`'${options.appPrefix}-${options.name}-list'`);

      const facadeSpecFile = tree.get(`/test/${pluralize(options.name)}-list/list.facade.spec.ts`);
      content = facadeSpecFile?.content.toString();
      expect(content).toContain(`[${classify(options.name)}Properties.ADDRESS_LINE_1]: 'ADDRESS_LINE_1',`);
      expect(content).toContain(`[${classify(options.name)}Properties.ADDRESS_LINE_1]: 'ADDRESS_LINE_1',`);
      done();
    }
    catch (err) {
      done.fail(err);
    }
  });
});
