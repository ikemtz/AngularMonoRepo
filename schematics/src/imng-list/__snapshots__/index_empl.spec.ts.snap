// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`imng-list generation works 1`] = `
"import { ODataState } from 'imng-kendo-odata';

export const employeeGridState: ODataState = {
  take: 20,
  skip: 0,
  selectors: [
    EmployeeProperties.ID, 
    EmployeeProperties.LAST_NAME, 
    EmployeeProperties.FIRST_NAME, 
    EmployeeProperties.BIRTH_DATE, 
    EmployeeProperties.MOBILE_PHONE, 
    EmployeeProperties.HOME_PHONE, 
    EmployeeProperties.PHOTO, 
    EmployeeProperties.EMAIL, 
    EmployeeProperties.ADDRESS_LINE_1, 
    EmployeeProperties.ADDRESS_LINE_2, 
    EmployeeProperties.CITY, 
    EmployeeProperties.STATE, 
    EmployeeProperties.ZIP, 
    EmployeeProperties.IS_ENABLED, 
    EmployeeProperties.HIRE_DATE, 
    EmployeeProperties.FIRE_DATE, 
    EmployeeProperties.TOTAL_HOURS_OF_SERVICE, 
    EmployeeProperties.CERTIFICATION_COUNT, 
    EmployeeProperties.COMPETENCY_COUNT, 
    EmployeeProperties.HEALTH_ITEM_COUNT, 
    EmployeeProperties.CREATED_BY, 
    EmployeeProperties.UPDATED_BY, 
    EmployeeProperties.CREATED_ON_UTC, 
    EmployeeProperties.UPDATED_ON_UTC, 
  ],
  sort: [
    { field: EmployeeProperties.LAST_NAME, dir: 'asc' },
  ], 
};"
`;

exports[`imng-list generation works: list.component.html 1`] = `
"<kendo-grid (detailExpand)="detailExpanded($event)" [height]="800" kendoGridFocusable [imngODataGrid]="this">
  <ng-template kendoGridToolbarTemplate>
    <div class="w-100 d-flex justify-content-between">
      <h3>Employees</h3>
      <imng-kendo-odata-grid-header
        entityName="Employee"
        (addItemClicked)="addItem()"
        (reloadEntitiesClicked)="reloadEntities()"
        (resetFiltersClicked)="resetFilters()"
        [hasHiddenColumns$]="hasHiddenColumns$" />
    </div>
  </ng-template>
  <kendo-grid-column [width]="75">
    <ng-template kendoGridCellTemplate let-dataItem let-rowIndex="rowIndex">
      <kendo-menu title="Actions">
        <kendo-menu-item [svgIcon]="icons.menu">
          <kendo-menu-item>
            <ng-template kendoMenuItemLinkTemplate>
              <button
                type="button"
                class="btn btn-sm mx-3 my-2 text-start"
                (click)="editItem(dataItem)"
                placement="left"
                tooltip="Edit">
                <kendo-svgicon [icon]="icons.menu" size="xlarge"></kendo-svgicon>
                Edit
              </button>
            </ng-template>
          </kendo-menu-item>
          <kendo-menu-item>
            <ng-template kendoMenuItemLinkTemplate>
              <button
                type="button"
                class="btn btn-sm mx-3 my-2 text-start"
                (click)="deleteItem(dataItem)"
                placement="left"
                tooltip="Delete">
                <kendo-svgicon [icon]="icons.trash" size="xlarge"></kendo-svgicon>
                Delete
              </button>
            </ng-template>
          </kendo-menu-item>
        </kendo-menu-item>
      </kendo-menu>
    </ng-template>
  </kendo-grid-column>

  <kendo-grid-column [field]="props.ID" [width]="100" title="Id" filter="text" [hidden]="false">
    <ng-template kendoGridCellTemplate let-dataItem>
      <imng-kendo-copy
        [displayValue]="\`\${(dataItem[props.ID] | slice: 0 : 4)}.-.\${(dataItem[props.ID] | slice: -4)}\`"
        [copyValue]="dataItem[props.ID]" />
    </ng-template>
    <ng-template kendoGridFilterMenuTemplate let-filter>
      <imng-uuid-filter [filter]="filter" />
    </ng-template>
    <ng-template class="header" kendoGridHeaderTemplate let-column>
      <span class="h6" title="Id">Id</span>
    </ng-template>
  </kendo-grid-column>
  <kendo-grid-column [field]="props.LAST_NAME" [width]="100" title="Last Name" filter="text" [hidden]="false">
    <ng-template class="header" kendoGridHeaderTemplate let-column>
      <span class="h6" title="Last Name">Last Name</span>
    </ng-template>
  </kendo-grid-column>
  <kendo-grid-column [field]="props.FIRST_NAME" [width]="100" title="First Name" filter="text" [hidden]="false">
    <ng-template class="header" kendoGridHeaderTemplate let-column>
      <span class="h6" title="First Name">First Name</span>
    </ng-template>
  </kendo-grid-column>
  <kendo-grid-column [field]="props.BIRTH_DATE" [width]="100" format="{0:MM/dd/yy}" title="Birth Date" filter="date" [hidden]="false">
    <ng-template class="header" kendoGridHeaderTemplate let-column>
      <span class="h6" title="Birth Date">Birth Date</span>
    </ng-template>
  </kendo-grid-column>
  <kendo-grid-column [field]="props.MOBILE_PHONE" [width]="100" title="Mobile Phone" filter="text" [hidden]="false">
    <ng-template class="header" kendoGridHeaderTemplate let-column>
      <span class="h6" title="Mobile Phone">Mobile Phone</span>
    </ng-template>
  </kendo-grid-column>
  <kendo-grid-column [field]="props.HOME_PHONE" [width]="100" title="Home Phone" filter="text" [hidden]="false">
    <ng-template class="header" kendoGridHeaderTemplate let-column>
      <span class="h6" title="Home Phone">Home Phone</span>
    </ng-template>
  </kendo-grid-column>
  <kendo-grid-column [field]="props.PHOTO" [width]="100" title="Photo" filter="text" [hidden]="false">
    <ng-template class="header" kendoGridHeaderTemplate let-column>
      <span class="h6" title="Photo">Photo</span>
    </ng-template>
  </kendo-grid-column>
  <kendo-grid-column [field]="props.EMAIL" [width]="100" title="Email" filter="text" [hidden]="false">
    <ng-template class="header" kendoGridHeaderTemplate let-column>
      <span class="h6" title="Email">Email</span>
    </ng-template>
  </kendo-grid-column>
  <kendo-grid-column [field]="props.ADDRESS_LINE_1" [width]="100" title="Address Line 1" filter="text" [hidden]="false">
    <ng-template class="header" kendoGridHeaderTemplate let-column>
      <span class="h6" title="Address Line 1">Address Line 1</span>
    </ng-template>
  </kendo-grid-column>
  <kendo-grid-column [field]="props.ADDRESS_LINE_2" [width]="100" title="Address Line 2" filter="text" [hidden]="false">
    <ng-template class="header" kendoGridHeaderTemplate let-column>
      <span class="h6" title="Address Line 2">Address Line 2</span>
    </ng-template>
  </kendo-grid-column>
  <kendo-grid-column [field]="props.CITY" [width]="100" title="City" filter="text" [hidden]="false">
    <ng-template class="header" kendoGridHeaderTemplate let-column>
      <span class="h6" title="City">City</span>
    </ng-template>
  </kendo-grid-column>
  <kendo-grid-column [field]="props.STATE" [width]="100" title="State" filter="text" [hidden]="false">
    <ng-template class="header" kendoGridHeaderTemplate let-column>
      <span class="h6" title="State">State</span>
    </ng-template>
  </kendo-grid-column>
  <kendo-grid-column [field]="props.ZIP" [width]="100" title="Zip" filter="text" [hidden]="false">
    <ng-template class="header" kendoGridHeaderTemplate let-column>
      <span class="h6" title="Zip">Zip</span>
    </ng-template>
  </kendo-grid-column>
  <kendo-grid-column [field]="props.IS_ENABLED" [width]="100" title="Is Enabled" filter="boolean" [hidden]="false">
    <ng-template kendoGridCellTemplate let-dataItem>
     @if (dataItem[props.IS_ENABLED]) {
        <kendo-svgicon class="text-success font-weight-bolder h2" [icon]="icons.check" size="xlarge"></kendo-svgicon>
      } @else {
        <kendo-svgicon class="text-danger font-weight-bolder h2" [icon]="icons.cancel" size="xlarge"></kendo-svgicon>
      }
    </ng-template>
    <ng-template class="header" kendoGridHeaderTemplate let-column>
      <span class="h6" title="Is Enabled">Is Enabled</span>
    </ng-template>
  </kendo-grid-column>
  <kendo-grid-column [field]="props.HIRE_DATE" [width]="100" format="{0:MM/dd/yy}" title="Hire Date" filter="date" [hidden]="false">
    <ng-template class="header" kendoGridHeaderTemplate let-column>
      <span class="h6" title="Hire Date">Hire Date</span>
    </ng-template>
  </kendo-grid-column>
  <kendo-grid-column [field]="props.FIRE_DATE" [width]="100" format="{0:MM/dd/yy}" title="Fire Date" filter="date" [hidden]="false">
    <ng-template class="header" kendoGridHeaderTemplate let-column>
      <span class="h6" title="Fire Date">Fire Date</span>
    </ng-template>
  </kendo-grid-column>
  <kendo-grid-column [field]="props.TOTAL_HOURS_OF_SERVICE" [width]="100" title="Total Hours Of Service" filter="numeric" [hidden]="false">
    <ng-template class="header" kendoGridHeaderTemplate let-column>
      <span class="h6" title="Total Hours Of Service">Total Hours Of Service</span>
    </ng-template>
  </kendo-grid-column>
  <kendo-grid-column [field]="props.CERTIFICATION_COUNT" [width]="100" title="Certification Count" filter="numeric" [hidden]="false">
    <ng-template class="header" kendoGridHeaderTemplate let-column>
      <span class="h6" title="Certification Count">Certification Count</span>
    </ng-template>
  </kendo-grid-column>
  <kendo-grid-column [field]="props.COMPETENCY_COUNT" [width]="100" title="Competency Count" filter="numeric" [hidden]="false">
    <ng-template class="header" kendoGridHeaderTemplate let-column>
      <span class="h6" title="Competency Count">Competency Count</span>
    </ng-template>
  </kendo-grid-column>
  <kendo-grid-column [field]="props.HEALTH_ITEM_COUNT" [width]="100" title="Health Item Count" filter="numeric" [hidden]="false">
    <ng-template class="header" kendoGridHeaderTemplate let-column>
      <span class="h6" title="Health Item Count">Health Item Count</span>
    </ng-template>
  </kendo-grid-column>
  <kendo-grid-column [field]="props.CREATED_BY" [width]="100" title="Created By" filter="text" [hidden]="true">
    <ng-template class="header" kendoGridHeaderTemplate let-column>
      <span class="h6" title="Created By">Created By</span>
    </ng-template>
  </kendo-grid-column>
  <kendo-grid-column [field]="props.UPDATED_BY" [width]="100" title="Updated By" filter="text" [hidden]="true">
    <ng-template class="header" kendoGridHeaderTemplate let-column>
      <span class="h6" title="Updated By">Updated By</span>
    </ng-template>
  </kendo-grid-column>
  <kendo-grid-column [field]="props.CREATED_ON_UTC" [width]="100" format="{0:MM/dd/yy HH:mm}" title="Created On Utc" filter="date" [hidden]="true">
    <ng-template class="header" kendoGridHeaderTemplate let-column>
      <span class="h6" title="Created On Utc">Created On Utc</span>
    </ng-template>
  </kendo-grid-column>
  <kendo-grid-column [field]="props.UPDATED_ON_UTC" [width]="100" format="{0:MM/dd/yy HH:mm}" title="Updated On Utc" filter="date" [hidden]="true">
    <ng-template class="header" kendoGridHeaderTemplate let-column>
      <span class="h6" title="Updated On Utc">Updated On Utc</span>
    </ng-template>
  </kendo-grid-column>

  <kendo-grid-pdf [fileName]="getExportFileName('Employees')" [allPages]="false">
    <kendo-grid-pdf-margin top="1cm" left="1cm" right="1cm" bottom="1cm" />
  </kendo-grid-pdf>
  <kendo-grid-excel [fileName]="getExportFileName('Employees')" [fetchData]="excelData" />
</kendo-grid>

@switch (crudFacade.currentModalState$ | async) {
  @case (modalStates.ADD) {
    <nrcrn-employee-add />
  }
  @case (modalStates.EDIT) {
    <nrcrn-employee-edit />
  }
  @case (modalStates.DELETE) {
    <imng-data-delete-dialog
      [facade]="crudFacade"
      [dataItem]="crudFacade.currentEntity$ | async">
      <div class="text-center">Are you sure you want to delete this record?</div>
    </imng-data-delete-dialog>
  }
}"
`;
